### PR TITLE
[libc] newheadergen: configured cmake

### PR DESCRIPTION
**Pull request by aaryanshukla**
- all headers in the build system are generated by newheadergen
- tested on gpu-build

### CODE REVIEW

**Comment by RoseZhang03** [comment]
LGTM!

File: libc/cmake/modules/LLVMLibCHeaderRules.cmake [hunk 1]
```diff
@@ -144,18 +243,18 @@ function(add_gen_header target_name)
     file(MAKE_DIRECTORY ${LIBC_INCLUDE_DIR}/llvm-libc-decls/gpu)
     set(decl_out_file ${LIBC_INCLUDE_DIR}/llvm-libc-decls/${relative_path})
     add_custom_command(
-      OUTPUT ${decl_out_file}
-      COMMAND ${hdrgen_exe} -o ${decl_out_file}
-              --header ${ADD_GEN_HDR_GEN_HDR} --def ${in_file} --export-decls
-              ${replacement_params} -I ${LIBC_SOURCE_DIR} ${ENTRYPOINT_NAME_LIST_ARG}
-              ${LIBC_SOURCE_DIR}/config/${LIBC_TARGET_OS}/api.td
+    OUTPUT ${decl_out_file}
```
Comment (michaelrj-google) [inline libc/cmake/modules/LLVMLibCHeaderRules.cmake:246]: why is this changed?

File: libc/include/CMakeLists.txt [hunk 2]
```diff
@@ -17,676 +17,1301 @@ add_header(
...
-  YAML_FILE ../libc/newhdrgen/yaml/dirent.yaml
-  DEF_FILE dirent.h.def
-  GEN_HDR dirent.h
-  DEPENDS
-    .llvm_libc_common_h
-    .llvm-libc-types.ino_t
-    .llvm-libc-types.DIR
-    .llvm-libc-types.struct_dirent
-)
+if (USE_NEW_HEADER_GEN)
```
Comment (aaryanshukla) [inline libc/include/CMakeLists.txt:20]: You may want to add the option command that can switch either on and off depending on how the user wants to build for now. Something like option(USE_NEW_HEADER_GEN OFF) before this line and then define a macro such as macro(new_function_name...) if...

File: libc/include/CMakeLists.txt [hunk 3]
```diff
@@ -17,676 +17,1301 @@ add_header(
...
+  add_gen_header2(
```
Comment (aaryanshukla) [inline libc/include/CMakeLists.txt:21]: add an else statement for the add_gen_header, old headergen.

File: libc/include/CMakeLists.txt [hunk 4]
```diff
@@ -17,627 +17,1301 @@ add_header(
...
-  dirent
-  DEF_FILE dirent.h.def
-  GEN_HDR dirent.h
-  DEPENDS
-    .llvm_libc_common_h
-    .llvm-libc-types.ino_t
-    .llvm-libc-types.DIR
-    .llvm-libc-types.struct_dirent
-)
+  add_gen_header2(
```
Comment (aaryanshukla) [inline libc/include/CMakeLists.txt:30]: make sure all of these are switched to the new function name

File: libc/cmake/modules/LLVMLibCHeaderRules.cmake (hunk 1 above)
Comment (RoseZhang03) [reply libc/cmake/modules/LLVMLibCHeaderRules.cmake:246]: I'm not too sure, that was there from Aaryan's commit. The code that was removed/replaced looks the same though.

Comment (aaryanshukla) [reply libc/cmake/modules/LLVMLibCHeaderRules.cmake:246]: Accidentally changed spacing when deleting and adding will make it the same again

File: libc/include/CMakeLists.txt (hunk 3 above)
Comment (RoseZhang03) [reply libc/include/CMakeLists.txt:21]: this is added at the very end

File: libc/include/CMakeLists.txt (hunk 2 above)
Comment (RoseZhang03) [reply libc/include/CMakeLists.txt]: option is added in libc/CMakesList.txt. Macro has now been added

File: libc/include/CMakeLists.txt [hunk 5]
```diff
@@ -17,1304 +17,697 @@ add_header(
...
-      .llvm-libc-types.mbstate_t
-      .llvm-libc-types.size_t
-      .llvm-libc-types.wint_t
-      .llvm-libc-types.wchar_t
-  )
-
-  if(LIBC_TARGET_OS_IS_GPU)
-    file(MAKE_DIRECTORY ${LIBC_INCLUDE_DIR}/gpu)
-
+macro(add_header_macro TYPE YAML_FILE DEF_FILE GEN_HDR DEPENDS)
```
Comment (aaryanshukla) [inline libc/include/CMakeLists.txt:20]: I think we can remove type since it is not defined in the libcheader

Comment (RoseZhang03) [reply libc/include/CMakeLists.txt:20]: Sorry, that was a misname. TYPE should be called NAME. It should used in both add_gen_header and add_gen_header2 right?

**Review by michaelrj-google — APPROVED** [review]
LGTM, thanks for adding the macro it makes it much easier to read.

**Review by aaryanshukla — COMMENTED** [review]
Not sure if my previous comment went through, but revert back the change for function.py and header.py. Since Gpu and Libc headers are dependent on function.py, and gpu headers do not have NO EXCEPT at the end, we can add NO EXCEPT to header.py instead.

**Review by michaelrj-google — APPROVED** [review]

**Comment by ilovepi** [comment]
Hi, we're seeing this break builds in our CI,, as well as in buildbot, can you revert your patch until you can fix the issue?
https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8742122933932616721/overview

**Comment by ilovepi** [comment]
Thank you! If you want to test your PR in our build we can probably help you vet it on our bots tomorrow.
